File: javascript/bob/bob.js
function isBlank(input){
  return !!input.match(/^(\s*)$/);
}
function isAllCaps(input){
  return input.toUpperCase() === input && !!input.match(/[A-Z]+/);
}
function isAQuestion(input){
  return !!input.match(/.+\?$/);
}

function Bob(){
  this.hey = function(input){
    if(isBlank(input)){ return "Fine. Be that way!"; } 
    else if(isAllCaps(input)){ return "Woah, chill out!"; } 
    else if(isAQuestion(input)){ return "Sure."; } 
    else { return "Whatever."; }
  }
}
module.exports = Bob;

File: ruby/bob/bob.rb
class Bob
  def hey(phrase)
    input = Input.new phrase
    
    case
    when input.blank?
      'Fine. Be that way!'
    when input.all_caps?
      'Woah, chill out!'
    when input.a_question?
      'Sure.'
    else
      'Whatever.'
    end
  end
end
class Input < String
  def blank?
    self.strip.empty?
  end
  def all_caps?
    !!self.match(/\A([^[a-z]]+)\z/) && !!self.match(/[A-Z]+/)
  end
  def a_question?
    self.end_with? '?'
  end
end